### DISCUSSION TITLE
Make it easier to specify the k/stiffness of a spring?

### DISCUSSION THREAD

**Discussion by salk31** [dyn4j/dyn4j#238] [Ideas]
Hi,

I'm enjoying the project and it is helping me work on my own. One tiny improvement that would let me dumb down my code and might be helpful to others is to be able to set the k/stiffness of a DistanceJoint.

What I've been using so far is a class that extends DistanceJoint and has the single method:
```java
    public void setK(double k) {
        double lm = this.getReducedMass();
        setFrequency(Math.sqrt(k / lm) / Geometry.TWO_PI);
    }
```

getReducedMass is protected to I couldn't see how to do this external to the class.

I was working on a pull request to get set/get K to DistanceJoint but belatedly saw that it already has a stiffness field so it might get messy. Also can getReducedMass change so breaking K?

Anyway, I thought it was better to start a discussion rather than diving in with my suggested code.

Cheers

Sam

**Comment by wnbittle** [comment]
The original idea was that you'd set the desired frequency using `setFrequency` so that if you changed the mass of either body, the frequency of oscillation would remain constant.

What you've defined here still sets the frequency, but by using an "arbitrary" constant based on the _current_ mass.  Since it still sets the frequency, if you changed the mass of the bodies, the frequency would remain constant as it does currently.  

However, the spring constant approach has two cons:

- If you asked for the K constant back in a `getK` method, you'd have to compute it, and it'd be based on the _current_ mass, and therefore may differ from what you set it to.
- Spring constants are hard to choose, especially in complex multi-body situations.

All that said, I kind of like the flexibility of having something like this.  What if it was something like this instead:
```java
// this would do what you have above
// maybe there's a better name?
public void setFrequencyUsingSpringConstant(double k);

// this would return the K value given the current state of the joint
public double getCurrentSpringConstant();
```

This would allow users to use spring constants if they wanted AND it would clearly convey that they are really setting the frequency.  Adding a getter would allow them to evaluate the spring constant as the mass of the bodies changes.

**Reply by wbpaley** [reply]
This would be useful to me, too. In general I love it where you track/implement/expose-methods-for the real physics in addition to the application-oriented niceties (like changing the constant to keep a certain behavior). 

I could easily see myself wanting to set up a constrained spring/mass system, then interactively change the masses--while assuming they're the same springs--to see how the behavior changes.

**Comment by salk31** [comment]
Thanks to both of you,

William, I like you method names much better than mine. Out of interest would getCurrentSpringConstant actually calculate it or just return DistanceJoint.stiffness? If so maybe use "Stiffness" rather than "SpringConstant" for consistency?

I did wonder about decomposing it so DistanceJoint pointed at a polymorphic SpringModel with three implementations Fixed/Stiffness/Frequency but overkill and even then not worth the pain of adding it so late in the day?

btw some way to get the current K is something that would be nice to have for my code.

Cheers

Sam

**Comment by salk31** [comment]
Honestly not trying to be annoying but for my real world project I'm experimenting with constant force springs. https://matmatch.com/learn/engineering/constant-force-spring-equations

DistanceJoint still seems to be the way to go. For now I'm just going to try and hack it by dynamically setting the frequency. Might be totally off topic here but if you were tempted by the decomposed/pluggable spring model.

Cheers

Sam

**Comment by wnbittle** [comment]
Sorry for the delay in response here - I'm playing around with a Spring definition that will allow either stiffness or frequency to be set.  There are some API decisions I need to make so it'll take me some time.

For constant force springs you may want to abandon the spring model entirely and just apply a constant force based on some custom logic each simulation iteration.  Another option is to set the stiffness high and clamp the total force of the spring (a feature I'm adding) to simulate a constant force spring.  What I'm struggling to understand is if a constant force spring applies the same constant force in both extension and compression.  If so, then clamping the force would achieve would achieve that goal.  Another thought is to allow the user to specify the desired force (then convert to impulse), but I suspect this will be very unstable (simulation wise).  The last thought is to create a more general spring model as you suggest, but this would require deep understanding of how dyn4j's global solving works (impulse-based, iterative).

**Comment by salk31** [comment]
Please don't worry about constant force springs or my suggestion about "spring model". I just thought they might help guide your design.

FWIW I don't think real world constant force springs don't really have compression. They are in spirals so once retracted (not stretched at all) the don't go any further.
